core/src/main/java/com/linecorp/armeria/common/metric/PrometheusRegistry.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.metric;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.internal.metric.PrometheusMetricRequestDecorator;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+
+/**
+ * Subclass {@link CollectorRegistry} that allows registering metrics with the same name.
+ */
+public class PrometheusRegistry extends CollectorRegistry {
+    private final ConcurrentMap<String, ? extends Collector> nameToCollector;
+
+    /**
+     * Creates a new instance.
+     */
+    public PrometheusRegistry() {
+        nameToCollector = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Registers a {@link Collector} conditionally.
+     * Registering the same name in the wrapped collector will throw an exception.
+     * So this wrapper provides register if-not-exists function.
+     * Allows use in multiple instances of {@link PrometheusMetricRequestDecorator}.
+     * @param name collector name
+     * @param createCollector function to generate a collector if absent
+     * @param <T> type of {@link Collector}
+     * @return existing or newly created collector
+     */
+    public <T extends Collector> T register(String name, Function<String, T> createCollector) {
+        ConcurrentMap<String, T> map = (ConcurrentMap<String, T>) nameToCollector;
+        Function<T, T> registerIfAbsent = collector -> {
+            super.register(collector);
```

Comment:
Could we add some code here that checks whether the `MetricFamilySamples.name` of the `collector` is equal to the specified `name`? e.g.

```java
List<Collector.MetricFamilySamples> mfs;
if (collector instanceof Collector.Describable) {
    mfs = ((Collector.Describable) collector).describe();
} else if (autoDescribe) {
    mfs = m.collect();
} else {
    throw new IllegalArgumentException("collector must have a name: " + collector);
}

if (mfs.size() != 1) {
    throw new IllegalArgumentException("collector must have a single name: " + collector);
}
```